mirai-core/src/commonMain/kotlin/net.mamoe.mirai/utils/MessageMatcher.kt

```diff
@@ -0,0 +1,442 @@
+}
+
+fun MessageSequence(messages: Iterable<SingleMessage>, reversed: Boolean = false): MessageSequence =
+    MessageSequence(messages.toList().toTypedArray(), reversed)
+
+/**
+ * 消息匹配类，通过多个 [Element] 来匹配 [MessageSequence]，支持正向和反向匹配。
+ *
+ * @see SingleMessage
+ * @see MessageMatcher.Element
+ **/
+class MessageMatcher(val elements: List<Element<*>>) {
+    constructor(vararg elements: Element<*>) : this(elements.toList())
+
+    abstract class Element<T : Any>(argumentName: String? = null) {
+        var argumentName: String? = argumentName
+            internal set
+        var matchedMessage: Message? = null
+            protected set
+
+        /**
+         * 从指定的 [messages] 消耗所需要的消息知道匹配成功，并加入新的元素（如果可能）。
+         * 如：消息为 ["你好啊！"]，匹配元素为"你好"。则匹配元素在消耗"你好啊！"这个消息后会加入一个"啊！"的消息，以备接下来的匹配。
+         *
+         * @return 是否匹配成功
+         */
+        abstract fun consume(messages: MessageSequence, resultMap: MutableMap<String, Any>?): Boolean
+
+        /**
+         * 用于子类设置解析结果
+         */
+        protected fun setResult(resultMap: MutableMap<String, Any>?, result: T) {
+            val argumentName = argumentName
+            if (resultMap != null && argumentName != null) resultMap[argumentName] = result
+        }
+
+        /**
+         * 用于子类设置解析结果（lazy）
+         */
+        protected fun setResult(resultMap: MutableMap<String, Any>?, result: Lazy<T>) {
+            val argumentName = argumentName
+            if (resultMap != null && argumentName != null) resultMap[argumentName] = result.value
+        }
+
+        @Suppress("unused")
+        protected inline fun <reified T : SingleMessage> Element<T>.getByType(messages: MessageSequence): T? {
+            val message = messages.first() ?: return null
+            if (message !is T) {
+                messages.put(message)
+                return null
+            }
+            return message
+        }
+
+        @Suppress("unused")
+        protected inline fun <reified T : SingleMessage> Element<T>.consumeByType(
+            messages: MessageSequence,
+            resultMap: MutableMap<String, Any>?
+        ): Boolean {
+            val message = getByType(messages) ?: return false
+            matchedMessage = message
+            setResult(resultMap, message)
+            return true
+        }
+
+        @Suppress("unused")
+        protected inline fun <reified T : SingleMessage> Element<T>.consumeByPredictAndType(
+            predict: (T) -> Boolean,
+            messages: MessageSequence,
+            resultMap: MutableMap<String, Any>?
+        ): Boolean {
+            val message = getByType(messages) ?: return false
+            return if (predict(message)) {
+                matchedMessage = message
+                setResult(resultMap, message)
+                true
+            } else false
+        }
+    }
+}
+
+fun <R> MessageMatcher(lambda: MessageMatcherBuilder.() -> R): MessageMatcher =
+    MessageMatcher(MessageMatcherBuilder().apply { lambda() }.messages)
```

Comment:
Use `apply(lambda)`

Reply:
> Use `apply(lambda)`

"apply" requires the return type to be Unit

Reply:
Unused generic type R is confusing. Replace with `Unit`